# Police Data Cleaning
## Crisis Data Dataset
### Exploration

Preview the data.

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)
df = pd.read_csv('../Datasets/Crisis_Data.csv')
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75082 entries, 0 to 75081
Data columns (total 25 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   Template ID                  75082 non-null  int64 
 1   Reported Date                75082 non-null  object
 2   Reported Time                75082 non-null  object
 3   Occurred Date / Time         75082 non-null  object
 4   Call Type                    75082 non-null  object
 5   Initial Call Type            75082 non-null  object
 6   Final Call Type              75082 non-null  object
 7   Disposition                  75082 non-null  object
 8   Use of Force Indicator       75082 non-null  object
 9   Subject Veteran Indicator    75082 non-null  object
 10  CIT Officer Requested        75082 non-null  object
 11  CIT Officer Dispatched       75082 non-null  object
 12  CIT Officer Arrived          75082 non-null  object
 13  Officer ID                   75

In [2]:
df.iloc[8:10]

Template ID        Reported Date Reported Time    Occurred Date / Time  \
8        57315  2015-05-15T00:00:00      08:30:00  05/15/2015 06:16:56 PM   
9        43946  2015-05-15T00:00:00      10:14:00  05/15/2015 06:47:32 PM   

  Call Type                                Initial Call Type  \
8       911  HAZ - POTENTIAL THRT TO PHYS SAFETY (NO HAZMAT)   
9       911     SUICIDE - IP/JO SUICIDAL PERSON AND ATTEMPTS   

                Final Call Type               Disposition  \
8         --DISTURBANCE - OTHER         Shelter Transport   
9  --CRISIS COMPLAINT - GENERAL  Emergent Detention / ITA   

  Use of Force Indicator Subject Veteran Indicator CIT Officer Requested  \
8                      N                         N                     N   
9                      N                         N                     Y   

  CIT Officer Dispatched CIT Officer Arrived Officer ID Officer Gender  \
8                      N                   Y       7666              F   
9                      Y                   Y       6115              M   

  Officer Race  Officer Year of Birth  Officer Years of Experience  \
8        White                   1970                            2   
9        White                   1968                           39   

  CIT Certified Indicator Officer Bureau Desc Officer Precinct Desc  \
8                       N   OPERATIONS BUREAU              EAST PCT   
9                       N   OPERATIONS BUREAU             NORTH PCT   

            Officer Squad Desc Precinct   Sector Beat  
8  EAST PCT 3RD W - E/G RELIEF        -        -    -  
9        NORTH PCT 2ND W - BOY    North  LINCOLN   L2

### Dealing with Missing Values
To make the missing data here more consistent, let's replace the dash values with None.

In [3]:
df.replace({'-': None}, inplace=True)
df['Occurred Date / Time'].replace({'1900-01-01': None}, inplace=True)


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75082 entries, 0 to 75081
Data columns (total 25 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   Template ID                  75082 non-null  int64 
 1   Reported Date                75082 non-null  object
 2   Reported Time                75082 non-null  object
 3   Occurred Date / Time         75082 non-null  object
 4   Call Type                    69482 non-null  object
 5   Initial Call Type            69482 non-null  object
 6   Final Call Type              69482 non-null  object
 7   Disposition                  73463 non-null  object
 8   Use of Force Indicator       75082 non-null  object
 9   Subject Veteran Indicator    75081 non-null  object
 10  CIT Officer Requested        75082 non-null  object
 11  CIT Officer Dispatched       75082 non-null  object
 12  CIT Officer Arrived          75082 non-null  object
 13  Officer ID                   75

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75082 entries, 0 to 75081
Data columns (total 25 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   Template ID                  75082 non-null  int64 
 1   Reported Date                75082 non-null  object
 2   Reported Time                75082 non-null  object
 3   Occurred Date / Time         75082 non-null  object
 4   Call Type                    69482 non-null  object
 5   Initial Call Type            69482 non-null  object
 6   Final Call Type              69482 non-null  object
 7   Disposition                  73463 non-null  object
 8   Use of Force Indicator       75082 non-null  object
 9   Subject Veteran Indicator    75081 non-null  object
 10  CIT Officer Requested        75082 non-null  object
 11  CIT Officer Dispatched       75082 non-null  object
 12  CIT Officer Arrived          75082 non-null  object
 13  Officer ID                   75

In [6]:
df.shape[0] - df.dropna().shape[0]

10328

There exist 10,328 rows with at least some missing data. 

In [7]:
df[(df['Officer Years of Experience'] == -1)].shape[0]


1056

In [8]:
df[(df['Officer Year of Birth'] == 1900)].shape[0]


377

There still exist some missing data in the integer columns in the form of officers being born in the year 1900, or officers having -1 years of experience. 


Let's look at optimizing the date columns.

In [9]:
df['Reported Date'].value_counts()['1900-01-01T00:00:00']

6

In [10]:
df['Occurred Date / Time'].value_counts()['01/01/1900 12:00:00 AM']

5600

There are some dates that are listed as occurring on January First, 1900. This must be a placeholder for missing / unknown data. Out of the 750082 entries, 6 have unknown reported dates, and 5600 have occurred dates that are unknown. 

In [11]:
date_cols = ['Reported Date', 'Reported Time', 'Occurred Date / Time']
for col in date_cols:
    df[col] = pd.to_datetime(df[col])


In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75082 entries, 0 to 75081
Data columns (total 25 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   Template ID                  75082 non-null  int64         
 1   Reported Date                75082 non-null  datetime64[ns]
 2   Reported Time                75082 non-null  datetime64[ns]
 3   Occurred Date / Time         75082 non-null  datetime64[ns]
 4   Call Type                    69482 non-null  object        
 5   Initial Call Type            69482 non-null  object        
 6   Final Call Type              69482 non-null  object        
 7   Disposition                  73463 non-null  object        
 8   Use of Force Indicator       75082 non-null  object        
 9   Subject Veteran Indicator    75081 non-null  object        
 10  CIT Officer Requested        75082 non-null  object        
 11  CIT Officer Dispatched       75082 non-nu

### Optimizing Numeric Columns
We can cast some columns such as the Officer Year of Birth to integer datatype and downcast it to the optimal subtype. The Officer ID columns has missing data, so it will remain as an object datatype. 


In [20]:
for col in ['Template ID', 'Officer Year of Birth', 'Officer Years of Experience']:
    df[col] = df[col].astype('int')
    df[col] = pd.to_numeric(df[col], downcast='integer')
    print(df[col].dtype, col)                               

int64 Template ID
int16 Officer Year of Birth
int8 Officer Years of Experience


### Optimizing Object Columns
Many columns have string values that repeat. We can optimize the space needed for this dataset by converting the columns with < 50% unique rows into category datatypes. 

In [14]:
for col in df.select_dtypes(include=['object']):
    num_unique_values = len(df[col].unique())
    num_total_values = len(df[col])
    print(col, num_unique_values, num_total_values)
    if num_unique_values / num_total_values < 0.5:
        df[col] = df[col].astype('category')

Call Type 9 75082
Initial Call Type 176 75082
Final Call Type 217 75082
Disposition 30 75082
Use of Force Indicator 2 75082
Subject Veteran Indicator 3 75082
CIT Officer Requested 2 75082
CIT Officer Dispatched 2 75082
CIT Officer Arrived 2 75082
Officer ID 1283 75082
Officer Gender 3 75082
Officer Race 9 75082
CIT Certified Indicator 2 75082
Officer Bureau Desc 7 75082
Officer Precinct Desc 27 75082
Officer Squad Desc 184 75082
Precinct 8 75082
Sector 18 75082
Beat 54 75082


In [16]:
df.iloc[5:15]

Template ID Reported Date       Reported Time Occurred Date / Time  \
5        552651    1900-01-01 2022-09-09 12:00:00  2019-04-18 01:02:18   
6         43479    2015-05-15 2022-09-09 11:21:00  2015-05-15 18:10:23   
7         43469    2015-05-15 2022-09-09 03:57:00  2015-05-15 11:16:25   
8         57315    2015-05-15 2022-09-09 08:30:00  2015-05-15 18:16:56   
9         43946    2015-05-15 2022-09-09 10:14:00  2015-05-15 18:47:32   
10        43653    2015-05-15 2022-09-09 08:33:00  2015-05-15 19:48:18   
11        43662    2015-05-15 2022-09-09 10:22:00  2015-05-15 21:33:28   
12        43992    2015-05-15 2022-09-09 10:56:00  2015-05-15 15:52:13   
13        44026    2015-05-16 2022-09-09 03:51:00  2015-05-16 13:57:12   
14        43929    2015-05-16 2022-09-09 06:20:00  2015-05-16 17:29:10   

   Call Type                                Initial Call Type  \
5        911                 DISTURBANCE, MISCELLANEOUS/OTHER   
6        911                 DISTURBANCE, MISCELLANEOUS/OTHER   
7        911            PERSON IN BEHAVIORAL/EMOTIONAL CRISIS   
8        911  HAZ - POTENTIAL THRT TO PHYS SAFETY (NO HAZMAT)   
9        911     SUICIDE - IP/JO SUICIDAL PERSON AND ATTEMPTS   
10       911            PERSON IN BEHAVIORAL/EMOTIONAL CRISIS   
11    ONVIEW  HAZ - POTENTIAL THRT TO PHYS SAFETY (NO HAZMAT)   
12       911        THEFT (DOES NOT INCLUDE SHOPLIFT OR SVCS)   
13       911                 DIST - IP/JO - DV DIST - NO ASLT   
14       911                          SERVICE - WELFARE CHECK   

                 Final Call Type  \
5   --CRISIS COMPLAINT - GENERAL   
6   --CRISIS COMPLAINT - GENERAL   
7   --CRISIS COMPLAINT - GENERAL   
8          --DISTURBANCE - OTHER   
9   --CRISIS COMPLAINT - GENERAL   
10  --CRISIS COMPLAINT - GENERAL   
11  --CRISIS COMPLAINT - GENERAL   
12           --THEFT - ALL OTHER   
13      --DISTURBANCE - JUVENILE   
14  --CRISIS COMPLAINT - GENERAL   

                                      Disposition Use of Force Indicator  \
5                              Resources Declined                      N   
6                        Emergent Detention / ITA                      N   
7                              Mobile Crisis Team                      N   
8                               Shelter Transport                      N   
9                        Emergent Detention / ITA                      N   
10                 No Action Possible / Necessary                      N   
11                 No Action Possible / Necessary                      N   
12             Geriatric Regional Assessment Team                      N   
13  Mental Health Agency or Case Manager Notified                      N   
14                 No Action Possible / Necessary                      N   

   Subject Veteran Indicator CIT Officer Requested CIT Officer Dispatched  \
5                          N                     N                      N   
6                          N                     N                      N   
7                          N                     Y                      Y   
8                          N                     N                      N   
9                          N                     Y                      Y   
10                         N                     N                      N   
11                         N                     N                      N   
12                         N                     N                      N   
13                         N                     N                      N   
14                         N                     N                      N   

   CIT Officer Arrived Officer ID Officer Gender               Officer Race  \
5                    Y       7490              M                      White   
6                    Y       7754              M                      White   
7                    N       7474              F                      White   
8                    Y       7666              F                      White

## Condensing Cleaning
Knowing what we do now, many of these operations can be consolidated and run during reading of the csv file. 

In [18]:
dtypes = {'Officer Year of Birth': 'int16', 'Officer Years of Experience': 'int8', 'Call Type': 'category'}
date_cols = ['Reported Date', 'Reported Time', 'Occurred Date / Time']

df = pd.read_csv('../Datasets/Crisis_Data.csv', dtype=dtypes, parse_dates=date_cols)
# df.info(memory_usage='deep')
df.iloc[5:20]

Template ID Reported Date       Reported Time Occurred Date / Time  \
5        552651    1900-01-01 2022-09-09 12:00:00  2019-04-18 01:02:18   
6         43479    2015-05-15 2022-09-09 11:21:00  2015-05-15 18:10:23   
7         43469    2015-05-15 2022-09-09 03:57:00  2015-05-15 11:16:25   
8         57315    2015-05-15 2022-09-09 08:30:00  2015-05-15 18:16:56   
9         43946    2015-05-15 2022-09-09 10:14:00  2015-05-15 18:47:32   
10        43653    2015-05-15 2022-09-09 08:33:00  2015-05-15 19:48:18   
11        43662    2015-05-15 2022-09-09 10:22:00  2015-05-15 21:33:28   
12        43992    2015-05-15 2022-09-09 10:56:00  2015-05-15 15:52:13   
13        44026    2015-05-16 2022-09-09 03:51:00  2015-05-16 13:57:12   
14        43929    2015-05-16 2022-09-09 06:20:00  2015-05-16 17:29:10   
15        44102    2015-05-16 2022-09-09 11:30:00  1900-01-01 00:00:00   
16        43982    2015-05-16 2022-09-09 11:07:00  2015-05-16 22:50:33   
17        43719    2015-05-16 2022-09-09 05:58:00  2015-05-16 03:15:18   
18        43832    2015-05-16 2022-09-09 01:24:00  2015-05-16 10:14:07   
19        43897    2015-05-16 2022-09-09 03:52:00  1900-01-01 00:00:00   

                   Call Type                                Initial Call Type  \
5                        911                 DISTURBANCE, MISCELLANEOUS/OTHER   
6                        911                 DISTURBANCE, MISCELLANEOUS/OTHER   
7                        911            PERSON IN BEHAVIORAL/EMOTIONAL CRISIS   
8                        911  HAZ - POTENTIAL THRT TO PHYS SAFETY (NO HAZMAT)   
9                        911     SUICIDE - IP/JO SUICIDAL PERSON AND ATTEMPTS   
10                       911            PERSON IN BEHAVIORAL/EMOTIONAL CRISIS   
11                    ONVIEW  HAZ - POTENTIAL THRT TO PHYS SAFETY (NO HAZMAT)   
12                       911        THEFT (DOES NOT INCLUDE SHOPLIFT OR SVCS)   
13                       911                 DIST - IP/JO - DV DIST - NO ASLT   
14                       911                          SERVICE - WELFARE CHECK   
15                         -                                                -   
16  TELEPHONE OTHER, NOT 911                          SERVICE - WELFARE CHECK   
17  TELEPHONE OTHER, NOT 911   UNKNOWN - ANI/ALI - WRLS PHNS (INCL OPEN LINE)   
18                       911                 DISTURBANCE, MISCELLANEOUS/OTHER   
19                         -                                                -   

                 Final Call Type  \
5   --CRISIS COMPLAINT - GENERAL   
6   --CRISIS COMPLAINT - GENERAL   
7   --CRISIS COMPLAINT - GENERAL   
8          --DISTURBANCE - OTHER   
9   --CRISIS COMPLAINT - GENERAL   
10  --CRISIS COMPLAINT - GENERAL   
11  --CRISIS COMPLAINT - GENERAL   
12           --THEFT - ALL OTHER   
13      --DISTURBANCE - JUVENILE   
14  --CRISIS COMPLAINT - GENERAL   
15                             -   
16  --CRISIS COMPLAINT - GENERAL   
17  --CRISIS COMPLAINT - GENERAL   
18         --DISTURBANCE - OTHER   
19                             -   

                                      Disposition Use of Force Indicator  \
5                              Resources Declined                      N   
6                        Emergent Detention / ITA                      N   
7                              Mobile Crisis Team                      N   
8                               Shelter Transport                      N   
9                        Emergent Detention / ITA                      N   
10                 No Action Possible / Necessary                      N   
11                 No Action Possible / Necessary                      N   
12             Geriatric Regional Assessment Team                      N   
13  Mental Health Agency or Case Manager Notified                      N   
14                 No Action Possible / Necessary                      N   
15                            Voluntary Committal                      N   
16                       Emergent

In [15]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75082 entries, 0 to 75081
Data columns (total 25 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   Template ID                  75082 non-null  int64         
 1   Reported Date                75082 non-null  datetime64[ns]
 2   Reported Time                75082 non-null  datetime64[ns]
 3   Occurred Date / Time         75082 non-null  datetime64[ns]
 4   Call Type                    69482 non-null  category      
 5   Initial Call Type            69482 non-null  category      
 6   Final Call Type              69482 non-null  category      
 7   Disposition                  73463 non-null  category      
 8   Use of Force Indicator       75082 non-null  category      
 9   Subject Veteran Indicator    75081 non-null  category      
 10  CIT Officer Requested        75082 non-null  category      
 11  CIT Officer Dispatched       75082 non-nu

What percentage of crisis calls result in a CIT certified Officer responding per year?

There exist 51 beats in Seattle. Why does this dataset have 54 unique beats? Which beats are not present in the dataset?